File: 第二阶段/2.8-collections模块精讲/collections_demo.py
# Python collections 模块示例
# 对应文章：2.8 collections模块精讲
#
# C# 老兵的 Python 修炼手册：
# collections 是 Python 标准库的"精品集合店"，提供了一堆比内置数据结构更好用的容器。
# 相当于 C# 的 System.Collections.Generic + ConcurrentDictionary 的一些功能，
# 但 Python 的版本更"贴心"，很多常用模式都帮你封装好了。

from collections import Counter, defaultdict, OrderedDict, deque, namedtuple, ChainMap

# ================================================================
#                       1. Counter —— 计数器
# ================================================================

print("=" * 60)
print("               1. Counter —— 计数器中的瑞士军刀")
print("=" * 60)

# 基础用法：词频统计
words = ["apple", "banana", "apple", "cherry", "banana", "apple"]
word_count = Counter(words)
print(f"\n--- 基础计数 ---")
print(f"  词频统计: {word_count}")  # Counter({'apple': 3, 'banana': 2, 'cherry': 1})
print(f"  最常见的2个: {word_count.most_common(2)}")

# 计数器运算（Counter 支持 + - & | 等集合运算，比 C# 的 Dictionary 灵活多了）
c1 = Counter(a=3, b=1)
c2 = Counter(a=1, b=2)
print(f"\n--- 计数器运算 ---")
print(f"  相加: {c1 + c2}")  # Counter({'a': 4, 'b': 3})
print(f"  相减: {c1 - c2}")  # Counter({'a': 2})
print(f"  交集: {c1 & c2}")  # Counter({'a': 1, 'b': 1})
print(f"  并集: {c1 | c2}")  # Counter({'a': 3, 'b': 2})

# 实战1：字符频率统计
print(f"\n--- 实战1：字符频率统计 ---")
text = "hello world"
char_count = Counter(text)
print(f"  字符频率: {dict(char_count.most_common())}")
print(f"  最常出现的字符: {char_count.most_common(1)[0]}")
# 找出只出现一次的字符
singles = [char for char, count in char_count.items() if count == 1]
print(f"  只出现一次的字符: {singles}")

# 实战2：从列表创建 Counter 并批量更新
print(f"\n--- 实战2：批量更新 ---")
shopping = Counter()
shopping.update(["apple", "banana", "apple"])
shopping.update(["apple", "cherry", "cherry", "cherry"])
print(f"  购物车: {shopping}")  # Counter({'apple': 3, 'cherry': 3, 'banana': 1})

# 实战3：正负过滤 —— 找出高频词和低频词
print(f"\n--- 实战3：正负过滤 ---")
corpus = Counter({"the": 500, "is": 400, "a": 300, "apple": 5, "banana": 2, "python": 10})
# 正数保留，负数排除
stopwords = Counter({"the": 500, "is": 400, "a": 300})
important = corpus - stopwords  # 去掉常见停用词
print(f"  去掉停用词后: {dict(important)}")

# C# 对应: 没有内置的 Counter，需要自己用 LINQ GroupBy + ToDictionary 实现
#           但 C# 的 LINQ 更通用，Counter 只是计数这一个场景

# ================================================================
#                       2. defaultdict —— 带默认值的字典
# ================================================================

print("\n" + "=" * 60)
print("               2. defaultdict —— 懒人字典")
print("=" * 60)

# 按首字母分组（经典用法）
groups = defaultdict(list)
for name in ["Alice", "Bob", "Anna", "Charlie", "Bob"]:
    groups[name[0]].append(name)
print(f"\n--- 按首字母分组 ---")
print(f"  结果: {dict(groups)}")

# 嵌套字典（多层默认值）
tree = defaultdict(lambda: defaultdict(list))
tree["animal"]["mammal"].append("cat")
tree["animal"]["bird"].append("eagle")
print(f"\n--- 嵌套结构 ---")
print(f"  树形结构: {dict(tree)}")

# C# 对应: Dictionary 不存在 key 时需要手动检查
#           或用 GetOrAdd() 方法（ConcurrentDictionary 也有）
#           但 defaultdict 写起来更简洁

# ================================================================
#                       3. deque —— 双端队列
# ================================================================

print("\n" + "=" * 60)
print("               3. deque —— 双端队列")
print("=" * 60)

dq = deque([1, 2, 3])
dq.appendleft(0)    # 左侧添加
dq.append(4)        # 右侧添加
dq.popleft()         # 左侧弹出
print(f"\n--- 基础操作 ---")
print(f"  deque: {dq}")  # deque([1, 2, 3, 4])

# 固定长度（自动丢弃另一端，非常适合做"最近N条"的场景）
print(f"\n--- 固定长度队列（自动丢弃）---")
recent = deque(maxlen=3)
for i in range(5):
    recent.append(i)
    print(f"  添加 {i}: {list(recent)}")

# 线程安全性说明
print(f"\n--- 线程安全性 ---")
print("  Python 的 deque 是线程安全的！（append/pop 操作是原子的）")
print("  但注意: 复合操作（如先检查再添加）不是原子的")
print("  对比 C#: ConcurrentQueue<T> / ConcurrentStack<T>")
print("  C# 的 LinkedList 不是线程安全的，需要手动加锁")
print("  Python 的 deque 在 C# 中最接近的是 ConcurrentQueue（但 deque 支持双端操作）")

# 实战：用 deque 做滑动窗口
print(f"\n--- 实战：滑动窗口 ---")
data = [10, 20, 30, 40, 50, 60]
window_size = 3
window = deque(maxlen=window_size)
for item in data:
    window.append(item)
    if len(window) == window_size:
        print(f"  窗口 {list(window)}, 平均值: {sum(window)/len(window):.1f}")

# ================================================================
#                       4. OrderedDict —— 有序字典
# ================================================================

print("\n" + "=" * 60)
print("               4. OrderedDict —— 老牌有序字典")
print("=" * 60)

od = OrderedDict()
od["first"] = 1
od["second"] = 2
od["third"] = 3
od.move_to_end("first")  # 移到末尾
print(f"\n--- 基础操作 ---")
print(f"  OrderedDict: {list(od.items())}")

# OrderedDict 的额外优势
print(f"\n--- OrderedDict vs 普通 dict ---")
print("  Python 3.7+ 普通 dict 已经保序，OrderedDict 的额外优势:")
print("  1. move_to_end() —— 可以把元素移到开头或末尾")
print("  2. 相等性比较考虑顺序: {a:1,b:2} != {b:2,a:1}")
print("  3. popitem(last=True/False) —— 可以从两端弹出")
print("  C# 对应: Dictionary<K,V> 默认保序（.NET Core+）")
print("  额外功能用 SortedList/KVP 按键排序")

# ================================================================
#                       5. namedtuple —— 命名元组
# ================================================================

print("\n" + "=" * 60)
print("               5. namedtuple —— 轻量级不可变对象")
print("=" * 60)

Point = namedtuple("Point", ["x", "y"])
p = Point(3, 4)
print(f"\n--- 基础用法 ---")
print(f"  点: {p.x}, {p.y}, 距离原点: {(p.x**2 + p.y**2)**0.5:.2f}")

# 带默认值
User = namedtuple("User", ["name", "age", "email"], defaults=["unknown", 0, ""])
user = User("Alice")
print(f"  用户: {user}")

# _replace() 方法 —— 创建修改后的新对象（不可变更新）
print(f"\n--- _replace() 不可变更新 ---")
p2 = p._replace(x=10)  # 创建新对象，原对象不变
print(f"  原始点: {p}")
print(f"  替换 x 后: {p2}")
print(f"  原始点不变: {p.x} == 3")  # 原对象没有被修改！
print("  C# 对比: record 类型的 With 表达式类似: p with { X = 10 }")

# _asdict() 转字典
print(f"\n--- _asdict() 转字典 ---")
print(f"  Point 转字典: {p._asdict()}")
print(f"  C# 对比: 类似 record 的 ToDictionary() 或匿名类型转换")

# 实战：函数返回多个值
print(f"\n--- 实战：函数返回多个值 ---")
def get_user_stats(users):
    """返回用户统计数据"""
    Stats = namedtuple("Stats", ["total", "active", "avg_age"])
    total = len(users)
    active = sum(1 for u in users if u.get("active", False))
    avg_age = sum(u.get("age", 0) for u in users) / total if total else 0
    return Stats(total=total, active=active, avg_age=avg_age)

users = [{"name": "Alice", "age": 30, "active": True},
         {"name": "Bob", "age": 25, "active": True},
         {"name": "Charlie", "age": 35, "active": False}]
stats = get_user_stats(users)
print(f"  统计: 总数={stats.total}, 活跃={stats.active}, 平均年龄={stats.avg_age:.1f}")

# ================================================================
#                       6. ChainMap —— 链式映射
# ================================================================

print("\n" + "=" * 60)
print("               6. ChainMap —— 链式映射（配置管理神器）")
print("=" * 60)

defaults = {"color": "red", "user": "guest", "debug": False}
config = {"color": "blue", "debug": True}
override = {"user": "admin"}

config_map = ChainMap(override, config, defaults)
print(f"\n--- 基础用法（配置合并）---")
print(f"  合并配置: {dict(config_map)}")
print(f"  color: {config_map['color']}")  # blue (优先级: override > config > defaults)
print(f"  user: {config_map['user']}")    # admin

# 模拟作用域链（Python 解释器就是这样处理作用域的！）
print(f"\n--- 作用域链模拟 ---")
local_vars = {"x": 10}
global_vars = {"x": 20, "y": 30}
builtin_vars = {"x": 40, "z": 50}
scope = ChainMap(local_vars, global_vars, builtin_vars)
print(f"  作用域查找 x: {scope['x']}")  # 10（局部优先）
print(f"  作用域查找 y: {scope['y']}")  # 30（全局）
print(f"  作用域查找 z: {scope['z']}")  # 50（内置）

# new_child() 方法 —— 创建子映射（在头部插入新层）
print(f"\n--- new_child() 创建子作用域 ---")
child_scope = scope.new_child({"x": 99, "w": 100})
print(f"  子作用域查找 x: {child_scope['x']}")  # 99
print(f"  子作用域查找 w: {child_scope['w']}")  # 100
print(f"  原作用域查找 x: {scope['x']}")         # 10（没被影响）
print(f"  C# 对比: 类似嵌套的 Dictionary 查找，但 ChainMap 不复制数据")
print(f"  ChainMap 只是把多个 dict 串起来，查找时按顺序找第一个匹配的")

# new_child() 的典型用途：临时覆盖配置
print(f"\n--- 实战：临时配置覆盖 ---")
base_config = {"timeout": 30, "retries": 3, "debug": False}
# 临时在测试环境中覆盖一些配置
test_config = ChainMap({"debug": True, "retries": 0}, base_config)
print(f"  基础配置: {base_config}")
print(f"  测试环境: {dict(test_config)}")
print(f"  测试完毕，丢弃 test_config 即可恢复，不会污染 base_config！")

print("\n" + "=" * 60)
print("  总结: collections 模块 = Python 的「精品集合店」")
print("  Counter 让计数变得优雅，deque 让队列操作飞起")
print("  namedtuple 让你不用写 class 就能拥有不可变对象")
print("  ChainMap 让配置管理变得简单又安全")
print("=" * 60)

File: 第二阶段/2.8-collections模块精讲/CollectionsDemo.cs
// C# 集合类示例
// 对应文章：2.8 collections模块精讲
//
// C# 老兵的 Python 修炼手册：
// Python 的 collections 是"精品集合店"，C# 的 System.Collections.Generic + ConcurrentCollections
// 是"大型超市" —— 品类更全，但 Python 的每个单品都打磨得特别好用。
// C# 的强项是类型安全和并发支持，Python 的强项是简洁和"开箱即用"。

using System;
using System.Collections.Generic;
using System.Collections.Concurrent;
using System.Collections.Specialized;
using System.Linq;

public class CollectionsDemo
{
    static void Main()
    {
        Console.WriteLine("===== collections 模块精讲（C# 对照版）=====\n");

        // ================================================================
        // 1. Counter → LINQ GroupBy + ToDictionary
        // ================================================================
        Console.WriteLine("===== 1. Counter vs LINQ GroupBy =====\n");

        // 基础计数
        Console.WriteLine("--- 1.1 基础词频统计 ---");
        var words = new[] { "apple", "banana", "apple", "cherry", "banana", "apple" };
        // Python: Counter(words)
        // C#:     需要 GroupBy + ToDictionary 三行代码...
        var wordCount = words.GroupBy(w => w)
            .OrderByDescending(g => g.Count())
            .ToDictionary(g => g.Key, g => g.Count());
        foreach (var kv in wordCount)
            Console.WriteLine($"  {kv.Key}: {kv.Value}");
        Console.WriteLine($"  最常见的: {wordCount.First().Key} ({wordCount.First().Value}次)");

        // 字符频率统计
        Console.WriteLine("\n--- 1.2 字符频率统计 ---");
        var text = "hello world";
        var charCount = text.GroupBy(c => c)
            .OrderByDescending(g => g.Count())
            .ToDictionary(g => g.Key, g => g.Count());
        Console.Write($"  字符频率: ");
        foreach (var kv in charCount)
            Console.Write($"'{kv.Key}'={kv.Count} ");
        Console.WriteLine();

        // C# 的优势：可以用 ConcurrentDictionary 在多线程下安全计数
        Console.WriteLine("\n--- 1.3 线程安全计数（C# 的优势！）---");
        Console.WriteLine("  Python 的 Counter 不是线程安全的，多线程需要用 Lock");
        Console.WriteLine("  C# 有 ConcurrentDictionary，天然支持多线程安全计数:");
        Console.WriteLine("  var counter = new ConcurrentDictionary<string, int>();");
        Console.WriteLine("  counter.AddOrUpdate(word, 1, (_, count) => count + 1);");
        Console.WriteLine("  这是 C# 比 Python 强的地方！");

        // ================================================================
        // 2. defaultdict → Dictionary + TryGetValue / GetOrAdd
        // ================================================================
        Console.WriteLine("\n===== 2. defaultdict vs Dictionary =====\n");

        Console.WriteLine("--- 2.1 按首字母分组 ---");
        // Python: groups = defaultdict(list) —— 一行搞定
        // C#:     需要手动检查 key 是否存在
        var groups = new Dictionary<char, List<string>>();
        foreach (var name in new[] { "Alice", "Bob", "Anna", "Charlie", "Bob" })
        {
            var key = name[0];
            // C# 的方式1：手动检查（和 Python 的 defaultdict 最接近）
            if (!groups.ContainsKey(key))
                groups[key] = new List<string>();
            groups[key].Add(name);

            // C# 的方式2：用 GetOrAdd（ConcurrentDictionary 也有这个方法）
            // groups.GetOrAdd(key, _ => new List<string>()).Add(name);
        }
        foreach (var kv in groups)
            Console.WriteLine($"  {kv.Key}: [{string.Join(", ", kv.Value)}]");

        // ================================================================
        // 3. deque → LinkedList / ConcurrentQueue
        // ================================================================
        Console.WriteLine("\n===== 3. deque vs LinkedList =====\n");

        Console.WriteLine("--- 3.1 基础操作 ---");
        var dq = new LinkedList<int>(new[] { 1, 2, 3 });
        dq.AddFirst(0);    // 左侧添加
        dq.AddLast(4);     // 右侧添加
        dq.RemoveFirst();  // 左侧弹出
        Console.Write("  LinkedList: ");
        foreach (var item in dq) Console.Write($"{item} ");
        Console.WriteLine();

        // 固定长度队列
        Console.WriteLine("\n--- 3.2 固定长度队列（模拟 deque(maxlen=N)）---");
        var recent = new FixedQueue<int>(3);
        for (int i = 0; i < 5; i++)
        {
            recent.Enqueue(i);
            Console.WriteLine($"  添加 {i}: [{string.Join(", ", recent)}]");
        }

        // 线程安全性对比
        Console.WriteLine("\n--- 3.3 线程安全性对比 ---");
        Console.WriteLine("  Python deque:  线程安全（append/pop 是原子操作）");
        Console.WriteLine("  C# LinkedList: 不是线程安全的！需要手动加锁");
        Console.WriteLine("  C# 替代方案:   ConcurrentQueue（仅支持尾部操作）");
        Console.WriteLine("                 ConcurrentStack（仅支持头部操作）");
        Console.WriteLine("                 要实现双端线程安全队列，需要自己封装 ConcurrentBag + 锁");
        Console.WriteLine("  Python 在这点上反而更方便！");

        // ================================================================
        // 4. OrderedDict → Dictionary（已内置保序）
        // ================================================================
        Console.WriteLine("\n===== 4. OrderedDict vs Dictionary =====\n");

        Console.WriteLine("--- 4.1 保序字典 ---");
        Console.WriteLine("  Python 3.7+ 普通 dict 已经保序");
        Console.WriteLine("  C# Dictionary<K,V> 自动保持插入顺序（.NET Core+）");
        Console.WriteLine("  额外功能需用 SortedList/KVP 按键排序");

        // 演示 C# Dictionary 的保序性
        var orderedDict = new Dictionary<string, int>();
        orderedDict["first"] = 1;
        orderedDict["second"] = 2;
        orderedDict["third"] = 3;
        Console.Write("  C# Dictionary 保序: ");
        foreach (var kv in orderedDict)
            Console.Write($"{kv.Key}={kv.Value} ");
        Console.WriteLine();

        // ================================================================
        // 5. namedtuple → record / ValueTuple
        // ================================================================
        Console.WriteLine("\n===== 5. namedtuple vs record / ValueTuple =====\n");

        Console.WriteLine("--- 5.1 基础用法 ---");
        // Python: Point = namedtuple("Point", ["x", "y"])
        // C#:     用 ValueTuple 最接近，或用 record（C# 9+）获得不可变性
        var p = (X: 3, Y: 4);
        Console.WriteLine($"  点: {p.X}, {p.Y}, 距离: {Math.Sqrt(p.X * p.X + p.Y * p.Y):F2}");

        // 用 record 模拟 namedtuple（不可变 + 值相等性）
        Console.WriteLine("\n--- 5.2 用 record 模拟（更接近 namedtuple 的语义）---");
        var point = new PointRecord(3, 4);
        var point2 = point with { X = 10 };  // 模拟 _replace()
        Console.WriteLine($"  原始点: {point}");
        Console.WriteLine($"  替换 X 后: {point2}");
        Console.WriteLine($"  原始点不变: X = {point.X}");
        Console.WriteLine($"  C# 的 with 表达式 = Python 的 _replace()，都是不可变更新！");

        // ================================================================
        // 6. ChainMap → 自定义实现 / 配置合并
        // ================================================================
        Console.WriteLine("\n===== 6. ChainMap vs 手动合并 =====\n");

        Console.WriteLine("--- 6.1 配置合并 ---");
        var defaults = new Dictionary<string, object>
        {
            ["color"] = "red", ["user"] = "guest", ["debug"] = false
        };
        var config = new Dictionary<string, object>
        {
            ["color"] = "blue", ["debug"] = true
        };
        var configMap = new ChainMapSample(defaults, config);
        Console.WriteLine($"  color: {configMap["color"]}");
        Console.WriteLine($"  user: {configMap["user"]}");

        // 模拟作用域链
        Console.WriteLine("\n--- 6.2 作用域链模拟 ---");
        var localVars = new Dictionary<string, object> { ["x"] = 10 };
        var globalVars = new Dictionary<string, object> { ["x"] = 20, ["y"] = 30 };
        var builtinVars = new Dictionary<string, object> { ["x"] = 40, ["z"] = 50 };
        var scope = new ChainMapSample(localVars, globalVars, builtinVars);
        Console.WriteLine($"  作用域查找 x: {scope["x"]}");  // 10
        Console.WriteLine($"  作用域查找 y: {scope["y"]}");  // 30
        Console.WriteLine($"  作用域查找 z: {scope["z"]}");  // 40

        // C# 中类似 ChainMap 的实现思路
        Console.WriteLine("\n--- 6.3 C# 中的替代方案 ---");
        Console.WriteLine("  C# 没有内置的 ChainMap，但可以用以下方式:");
        Console.WriteLine("  1. 手动实现（如上面的 ChainMapSample）");
        Console.WriteLine("  2. 用 Dictionary 合并: defaults.Concat(config).ToDictionary(...)");
        Console.WriteLine("  3. 用 .NET 8+ 的 FrozenDictionary 做不可变快照");
        Console.WriteLine("  Python 的 ChainMap 不复制数据，只做查找代理，更省内存！");

        Console.WriteLine("\n===== 总结: Python 的 collections 更'贴心'，C# 的更'强壮' =====");
    }
}

// record 类型（模拟 Python 的 namedtuple）
// C# 9+ 语法，自带 ToString、值相等性、不可变性
public record PointRecord(double X, double Y);

// 固定长度队列（模拟 Python deque(maxlen=N)）
// Python 的 deque 内置支持 maxlen，C# 需要自己封装
public class FixedQueue<T> : Queue<T>
{
    private readonly int _maxSize;
    public FixedQueue(int maxSize) => _maxSize = maxSize;
    public new void Enqueue(T item)
    {
        if (Count >= _maxSize) Dequeue();
        base.Enqueue(item);
    }
    public override string ToString() => string.Join(", ", this);
}

// ChainMap 简化实现（模拟 Python 的 collections.ChainMap）
// Python 的 ChainMap 支持 new_child()、maps 属性等，这里只实现核心的查找功能
public class ChainMapSample
{
    private readonly List<Dictionary<string, object>> _maps;
    public ChainMapSample(params Dictionary<string, object>[] maps)
        => _maps = maps.ToList();

    public object this[string key]
    {
        get
        {
            foreach (var map in _maps)
                if (map.ContainsKey(key))
                    return map[key];
            throw new KeyNotFoundException($"Key '{key}' not found");
        }
    }
}
